tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,96 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
```

Comment:
I would have called these `callback` and `cookie` personally.  Not a big deal, but what does the verbosity get you?